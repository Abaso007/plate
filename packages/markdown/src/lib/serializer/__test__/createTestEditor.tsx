import { createSlateEditor } from '@udecode/plate';
import { BaseIndentListPlugin } from '@udecode/plate-indent-list';

import { markdownPlugins } from '../../../../../../apps/www/src/registry/default/components/editor/plugins/markdown-plugins';
import { SuggestionPlugin } from '@udecode/plate-suggestion/react';

export const createTestEditor = (plugins: any[] = []) =>
  createSlateEditor({
    plugins: [markdownPlugins, ...plugins, SuggestionPlugin],
  });
