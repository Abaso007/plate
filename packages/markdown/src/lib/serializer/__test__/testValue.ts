export const testValue = [
  {
    type: 'h3',
    children: [
      {
        text: 'Heading',
      },
    ],
  },
  {
    type: 'p',
    indent: 1,
    listStyleType: 'disc',
    children: [
      {
        text: '1',
      },
    ],
  },
  {
    type: 'p',
    indent: 2,
    listStyleType: 'disc',
    children: [
      {
        text: '2',
      },
    ],
  },
  {
    type: 'p',
    indent: 3,
    listStyleType: 'disc',
    children: [
      {
        text: '3',
      },
    ],
  },
  {
    type: 'p',
    indent: 4,
    listStyleType: 'disc',
    children: [
      {
        text: '4',
      },
    ],
  },
  {
    children: [
      {
        children: [
          {
            children: [
              {
                children: [
                  {
                    text: '1-1',
                  },
                ],
                type: 'p',
              },
            ],
            type: 'td',
          },
          {
            children: [
              {
                children: [
                  {
                    text: '1-2',
                  },
                ],
                type: 'p',
              },
            ],
            type: 'td',
          },
        ],
        type: 'tr',
      },
      {
        children: [
          {
            children: [
              {
                children: [
                  {
                    text: '2-1',
                  },
                ],
                type: 'p',
              },
            ],
            type: 'td',
          },
          {
            children: [
              {
                children: [
                  {
                    text: '2-2',
                  },
                ],
                type: 'p',
              },
            ],
            type: 'td',
          },
        ],
        type: 'tr',
      },
    ],
    type: 'table',
  },
  {
    children: [
      {
        text: 'blockQuote',
      },
    ],
    type: 'blockquote',
  },
  {
    children: [
      {
        text: '',
      },
    ],
    type: 'hr',
  },
  {
    children: [
      {
        text: 'normal ',
      },
      {
        bold: true,
        text: 'bold',
      },
      {
        italic: true,
        text: 'italic',
      },
      {
        text: 'strikethrough',
        strikethrough: true,
      },
      {
        text: 'code',
        code: true,
      },
      {
        text: 'combine',
        bold: true,
        italic: true,
      },
      {
        children: [{ text: 'link' }],
        type: 'a',
        url: 'https://example.com',
        target: '_blank',
      },
      {
        children: [
          {
            text: '',
          },
        ],
        texExpression: 'E=mc^2',
        type: 'inline_equation',
      },
    ],
    type: 'p',
  },
  {
    caption: [{ text: 'test' }],
    children: [{ text: '' }],
    type: 'img',
    url: 'https://images.unsplash.com/photo-1712688930249-98e1963af7bd?q=80&w=2070&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D',
  },
  {
    type: 'code_block',
    lang: 'javascript',
    children: [
      {
        type: 'code_line',
        children: [
          {
            text: '// Use code blocks to showcase code snippets',
          },
        ],
      },
      {
        type: 'code_line',
        children: [
          {
            text: 'function greet() {',
          },
        ],
      },
      {
        type: 'code_line',
        children: [
          {
            text: "  console.info('Hello World!');",
          },
        ],
      },
      {
        type: 'code_line',
        children: [
          {
            text: '}',
          },
        ],
      },
    ],
  },
  {
    children: [
      {
        text: '',
      },
    ],
    texExpression: 'f(x)',
    type: 'equation',
  },
];
