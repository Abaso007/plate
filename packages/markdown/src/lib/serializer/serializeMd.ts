import type { Descendant, SlateEditor } from '@udecode/plate';

import remarkStringify from 'remark-stringify';
import { type Plugin, unified } from 'unified';

import type * as mdast from '../mdast';

import { MarkdownPlugin } from '../MarkdownPlugin';
import { convertNodesSerialize } from './convertNodes';

export type SerializeMdOptions = {
  editor: SlateEditor;
};

/** Serialize the editor value to Markdown. */
export const serializeMd = (
  editor: SlateEditor,

  // TODO:
  options?: Omit<SerializeMdOptions, 'editor'> & {
    value?: Descendant[];
  }
) => {
  const remarkPlugins: Plugin[] =
    editor.getOptions(MarkdownPlugin).remarkPlugins;

  const toRemarkProcessor = unified().use(remarkPlugins).use(remarkStringify);

  const nodesToSerialize = options?.value ?? editor.children;

  return toRemarkProcessor.stringify(
    slateToMdast({
      nodes: nodesToSerialize,
      options: {
        editor,
      },
    })
  );
};

const slateToMdast = ({
  nodes,
  options,
}: {
  nodes: Descendant[];
  options: SerializeMdOptions;
}): mdast.Root => {
  return {
    children: convertNodesSerialize(nodes, options) as mdast.Root['children'],
    type: 'root',
  } as mdast.Root;
};
