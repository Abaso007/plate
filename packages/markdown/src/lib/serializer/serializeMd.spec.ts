import { createTestEditor } from './__tests__/createTestEditor';
import { testValue } from './__tests__/testValue';
import { serializeMd } from './serializeMd';

describe('serializeMd', () => {
  it('should serialize a simple paragraph', () => {
    const editor = createTestEditor();
    const result = serializeMd(editor as any, { value: testValue });
    expect(result).toMatchSnapshot();
  });
});
