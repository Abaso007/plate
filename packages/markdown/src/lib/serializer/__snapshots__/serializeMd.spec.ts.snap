// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeMd should serialize a simple paragraph 1`] = `
"# Playground

## Comments and suggestions

Add comments to your content to provide additional conte1xt, suggesions, or collaborate with others

### Heading

* 1

  * 2

    * 3

      * 4

| 1-1 | 1-2 |
| --- | --- |
| 2-1 | 2-2 |

> blockQuote

***

normal **bold***italic*~~strikethrough~~\`code\`***combine***[link](https://example.com)$E=mc^2$

![test](https://images.unsplash.com/photo-1712688930249-98e1963af7bd?q=80\\&w=2070\\&auto=format\\&fit=crop\\&ixlib=rb-4.0.3\\&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D "test")

\`\`\`javascript
// Use code blocks to showcase code snippets
function greet() {
  console.info('Hello World!');
}
\`\`\`

$$
f(x)
$$

* [x] Todo item

* Numbered list item

Centered text

Text with custom line height

Mention: User

Date: 2024-01-01

Text with highlight and underline and keyboard formatting

Text with color and background color
"
`;
